### DISCUSSION TITLE
$watch() is not working for a Set()

### DISCUSSION THREAD

**Discussion by simonferndriger** [alpinejs/alpine#3491] [7. Bugs (archived)]
This is annoying, since sets are a very handy way to manage filters for a list for example. Adding or removing a filter with `delete()` and `add()` is smooth, but doesn't help if the list does not get updated because the change is not recognized by `$watch()` 😕

**Comment by ekwoka** [comment]
If you want to do this, you should use either `x-effect` or `Alpine.effect` instead of `$watch`.

`$watch` subscribes by `JSON.stringify` the target to access all it's property to register effects. But this does not really work with a Set. I'm not sure if there would be a reasonable way to do that to allow sets without quite a bit of work.

but `effect` works to do whatever it is you need.

Most of where the set support in the proxy comes from is in the reactivity engine which is `@vue/reactivity` and not in the Alpine code.

**Reply by simonferndriger** [reply]
Thank you! What about performance? Can I use x-effect in the `<html x-effect="...">` element to watch everything or would that be slowig down the website?

**Reply by simonferndriger** [reply]
However, I'm afraid if I used x-effect, the connected api calls would be made too often. To avoid this, I had to check if the request actually changed which is more complicated than just using arrays instead of sets. Would be awesome if sets could be watched specifically.

**Reply by ekwoka** [reply]
Any performance concerns with `x-effect` would be the same with `$watch` realistically. So I'm not sure I can really answer that, since neither is really better than the other in that way.

If your action does too much stuff, that is a separate issue.
